test/e2e/network/netpol/kubemanager.go

```diff
@@ -185,18 +180,57 @@ func (k *kubeManager) probeConnectivity(args *probeConnectivityArgs) (bool, stri
 
 	attempt := 0
 
-	err := wait.PollImmediate(time.Duration(args.pollIntervalSeconds)*time.Second, time.Duration(args.pollTimeoutSeconds)*time.Second,
-		func() (bool, error) {
-			stdout, stderr, err := k.executeRemoteCommand(args.nsFrom, args.podFrom, args.containerFrom, cmd)
-			if err != nil {
-				framework.Logf("retrying probe #%d :: %s/%s -> %s: error when running command: err - %v /// stdout - %s /// stderr - %s", attempt+1, args.nsFrom, args.podFrom, args.addrTo, err, stdout, stderr)
+	// NOTE: The return value of this function[probeConnectivity] should be true if the probe is successful and false otherwise.
+
+	// probeError will be the return value of this function[probeConnectivity] call.
+	var probeError error
+	var stderr string
+
+	// Instead of re-running the job on connectivity failure, the following conditionFunc when passed to PollImmediate, reruns
+	// the job when the observed value don't match the expected value, so we don't rely on return value of PollImmediate, we
+	// simply discard it and use probeError, defined outside scope of conditionFunc, for returning the result of probeConnectivity.
+	conditionFunc := func() (bool, error) {
+		_, stderr, probeError = k.executeRemoteCommand(args.nsFrom, args.podFrom, args.containerFrom, cmd)
+		// retry should only occur if expected and observed value don't match.
+		if args.expectConnectivity {
+			if probeError != nil {
+				// since we expect connectivity here, we fail the condition for PollImmediate to reattempt the probe.
+				// this happens in the cases where network is congested, we don't have any policy rejecting traffic
+				// from "podFrom" to "podTo" and probes from "podFrom" to "podTo" are failing.
+				framework.Logf("probe #%d :: connectivity expected :: %s/%s -> %s :: stderr - %s",
+					attempt+1, args.nsFrom, args.podFrom, args.addrTo, stderr,
+				)
+				attempt++
+				return false, nil
+			} else {
+				// we got the expected results, exit immediately.
+				return true, nil
+			}
+		} else {
+			if probeError != nil {
+				// we got the expected results, exit immediately.
+				return true, nil
+			} else {
+				// since we don't expect connectivity here, we fail the condition for PollImmediate to reattempt the probe.
+				// this happens in the cases where we have policy rejecting traffic from "podFrom" to "podTo", but CNI takes
+				// time to implement the policy and probe from "podFrom" to "podTo" was successful in that window.
+				framework.Logf(" probe #%d :: connectivity not expected :: %s/%s -> %s",
+					attempt+1, args.nsFrom, args.podFrom, args.addrTo,
+				)
 				attempt++
 				return false, nil
 			}
-			return true, nil
-		})
+		}
+	}
 
-	if err != nil {
+	// ignore the result of PollImmediate, we are only concerned with probeError.
```

Comment:
but if you timeout how do you know what is the right result?

Reply:
@aojea In case of timeout I can return a negative of the expected state to mark the test as a failure, what do you think?

Reply:
Meanwhile you clearly comment it, is that or returning an additional error, I can't see any more options.

I let you decide, pick your poison 😜